src/dataflow/render/mod.rs

```diff
@@ -314,6 +319,61 @@ pub(crate) fn build_dataflow<A: Allocate>(
                             Envelope::Upsert(_) => unreachable!(),
                         };
 
+                        // Implement source filtering and projection.
+                        // At the moment this is strictly optional, but we perform it anyhow
+                        // to demonstrate the intended use.
+                        if let Some(operators) = src.operators.clone() {
+                            // Determine replacement values for unused columns.
+                            let source_type = src.desc.typ();
+                            let position_or = (0..source_type.arity())
+                                .map(|col| {
+                                    if operators.projection.contains(&col) {
+                                        Ok(col)
+                                    } else {
+                                        Err({
+                                            let typ = &source_type.column_types[col];
+                                            if typ.nullable {
+                                                Datum::Null
+                                            } else {
+                                                typ.scalar_type.dummy_datum()
+                                            }
+                                        })
+                                    }
+                                })
+                                .collect::<Vec<_>>();
+
+                            // Evaluate the predicate on each record, noting potential errors that might result.
+                            let (collection2, errors) =
+                                collection.flat_map_fallible(move |input_row| {
+                                    let temp_storage = RowArena::new();
```

Comment:
Can we avoid an allocation per row by creating this outside the closure and moving it in?

Reply:
No, unfortunately. We need to repeatedly *drop* `temp_storage`, otherwise it may grow without bound. It is a memory leak if any of the predicates need to allocate. It could be allocated per-batch of records, by using `unary_fallible`, but .. we have a lot of this through the code and we should probably do that consistently too (it makes the code much less readable to use `unary_fallible`).

Reply:
Could we make it so that `predicate.eval` truncates it to zero before using? So it would reuse the same backing allocation (unless capacity needs to grow).

Reply:
We would need to invoke @jamii to sort that out. It makes sense, but I think that method would have to be unsafe (if you called it while a datum "borrowing" from `temp_storage` was still around it would be unsafe).

Reply:
Yeah, it would be an unsafe function.

The old api did this by having two types - a reusable storage and a temporary owner of lifetimes. The datum lifetime was tied to the latter and once it was dropped the storage got zeroed.